### ISSUE TITLE
Implement "quantum Tanner transformation" for `HGPCode`s and `LPCode`s

### ISSUE THREAD

**Issue by perlinm** [Infleqtion/qLDPC#113]
See:
- https://arxiv.org/abs/2309.11719
- https://inria.hal.science/hal-04206478/document

**Comment by qodesign** [comment]
I think "quantum tanner transform" is at best loosely defined. 

The second paper gives a reference for it and it's nowhere defined there.
There is a concrete construction for the [[25,4,3]] code from an HGP [[34,4,3]] code in that paper and the algorithm to remove the "secondary qubits" is well described...even though the explicit stabilizers they give in equation A2 have typos.

Here's code that does the construction. It's in GAP which I'm more familiar with than python; I think it should be straight forward to translate it to python and generalize it (an AI might be able to do that too but I didn't try)


```
# helper function (iterated tensor product)
Ten:=mats->Iterated(mats,KroneckerProduct);;

# helper function (horizontal concatenation of matrices, ~ matlab horzcat)
HrzCat:=function(mats)local n,k,T,V,a,b;
 n:=Length(mats);k:=DimensionsMat(mats[1])[1];
 T:=[];for b in [1..k] do V:=[];for a in [1..n] do V:=Concatenation(V,mats[a][b]);od;Add(T,V);od;
 return T;end;;

# base code from paper
H:=[[1,1,0,0,0],[0,1,1,1,0],[0,0,0,1,1]];

# HGP construction
Ha:=H;ha:=TransposedMat(Ha);T:=DimensionsMat(Ha);ma:=T[1];na:=T[2];ra:=RankMat(Ha*Z(2));
Hb:=H;hb:=TransposedMat(Hb);T:=DimensionsMat(Hb);mb:=T[1];nb:=T[2];rb:=RankMat(Hb*Z(2));
n:=na*nb+ma*mb;
k:=(na-ra)*(nb-rb)+(ma-ra)*(mb-rb);
LftQubits:=[1..na*nb];
RgtQubits:=na*nb+[1..ma*mb];;

Hz:=HrzCat([Ten([Ha,IdentityMat(nb)]),Ten([IdentityMat(ma),hb])]);;
Hx:=HrzCat([Ten([IdentityMat(na),Hb]),Ten([ha,IdentityMat(mb)])]);;

Print("\n original HGP code \n");
Print("dim(Hx)=",DimensionsMat(Hx)," weights = ",Collected(List(Hx,Sum)),"\n");
Print("dim(Hz)=",DimensionsMat(Hz)," weights = ",Collected(List(Hz,Sum)),"\n");
# optional distance calculation if uncommented
#LoadPackage("QDistRnd","0");d:=DistRandCSS(Hx*Z(2),Hz*Z(2),100,0,2);Print("distance = ",d,"\n");

# remove the "secondary qubits"
Ind:=RgtQubits;
for m in [1..ma*mb] do
 c:=First(Ind);
 if(\mod(m,2)=0)then a:=First([1..Length(Hz)],x->Hz[x][c]=1);T:=[];for b in [1..Length(Hz)] do if(b<>a)then Add(T,Hz[b]+Hz[b][c]*Hz[a]);fi;od;Hz:=T;fi;;
 if(\mod(m,2)=1)then a:=First([1..Length(Hx)],x->Hx[x][c]=1);T:=[];for b in [1..Length(Hx)] do if(b<>a)then Add(T,Hx[b]+Hx[b][c]*Hx[a]);fi;od;Hx:=T;fi;;
 Ind:=Ind{[2..Length(Ind)]};;
 Hx:=Hx{[1..Length(Hx)]}{Concatenation(LftQubits,Ind)};;
 Hz:=Hz{[1..Length(Hz)]}{Concatenation(LftQubits,Ind)};;
 Ind:=Ind-1;
od;;

Print("\n reduced code \n");
Print("dim(Hx)=",DimensionsMat(Hx)," weights = ",Collected(List(Hx,Sum)),"\n");
Print("dim(Hz)=",DimensionsMat(Hz)," weights = ",Collected(List(Hz,Sum)),"\n");
# optional distance calculation if uncommented
#LoadPackage("QDistRnd","0");d:=DistRandCSS(Hx*Z(2),Hz*Z(2),100,0,2);Print("distance = ",d,"\n");
```

if you read in the file with the distance calculation uncommented you should get : 

```
 original HGP code 
dim(Hx)=[ 15, 34 ] weights = [ [ 3, 6 ], [ 4, 7 ], [ 5, 2 ] ]
dim(Hz)=[ 15, 34 ] weights = [ [ 3, 6 ], [ 4, 7 ], [ 5, 2 ] ]
distance = 3

 reduced code 
dim(Hx)=[ 10, 25 ] weights = [ [ 2, 2 ], [ 3, 2 ], [ 4, 4 ], [ 6, 2 ] ]
dim(Hz)=[ 11, 25 ] weights = [ [ 2, 4 ], [ 3, 1 ], [ 4, 4 ], [ 6, 2 ] ]
distance = 3
```

**Comment by perlinm** [comment]
I am a bit confused.  Is your GAP just implementing this transformation for a specific HGPCode?  The issue is for an *arbitrary* HGPCode.

I thought about this issue a while ago at this point, but I think I remember being convinced that the transformation is well (albeit poorly) defined in arxiv:2206.07571.  I never figured out how to implement a general form of the transformation without having to solve an integer linear program, though.

**Comment by qodesign** [comment]
I think it should work for any HGP code...i don't if it's an exact match for "quantum tanner transform" but it removes all secondary qubits.  These are well defined for all HGP codes. Actually at any stage of the construction you get a legitimate code with fewer qubits (and same k and d).

**Comment by qodesign** [comment]
to get toric code of distance 5, replace H by : 
L:=5;H:=NullMat(L-1,L);for k in [1..L-1] do H[k][k]:=1;H[k][k+1]:=1;od;H:=Concatenation(H,[\mod(Sum(H),2)]);
this produced a [[25,2,4]] code from [[50,2,5]]; n and k look right but d should have stayed 5...I'll look into that laer

to get surface code of distance 5 replace H by
L:=5;H:=NullMat(L-1,L);for k in [1..L-1] do H[k][k]:=1;H[k][k+1]:=1;od;
this produced a [[25,1,25]] code from [[41,1,5]]; this looks right including the distance

**Comment by perlinm** [comment]
To be honest I can't parse the GAP code.  As a sanity check: does your quantum tanner transformation allow for something like
```python
from qldpc import codes
subcode = codes.HammingCode(3)  # classical [7, 4, 3] code
code = codes.HGPCode(subcode, subcode)  # quantum [58, 16, d] code

reduced_code = quantum_tanner_transformation(code)  # [49, 16, d] code
```

**Comment by qodesign** [comment]
yes it does. I edited the code with H coming from the [7,4,3] code. I used the exact matrix returned by your python code.
It does produce a [[58,16,3]] code and reduce it to [[49,16,3]]. Here's the GAP output : 

```
 original HGP code
dim(Hx)=[ 21, 58 ] weights = [ [ 5, 9 ], [ 6, 9 ], [ 7, 3 ] ]
dim(Hz)=[ 21, 58 ] weights = [ [ 5, 9 ], [ 6, 9 ], [ 7, 3 ] ]
distance = 3

 reduced code
dim(Hx)=[ 16, 49 ] weights = [ [ 4, 5 ], [ 8, 7 ], [ 12, 4 ] ]
dim(Hz)=[ 17, 49 ] weights = [ [ 4, 7 ], [ 8, 8 ], [ 12, 2 ] ]
distance = 3
```
note that m=21+21 for the HGP code, so k=58-42=16
and m=16+17 for the reduced code, so k=49-33=16

**Comment by perlinm** [comment]
Amazing.  Any interest in opening a pull request to add this to `qldpc`?  🙂

AND/OR: could you explain in words what exactly is happening in the transformation?  It looks like the crux is making sense of
```gap
if(\mod(m,2)=0)then a:=First([1..Length(Hz)],x->Hz[x][c]=1);if(a<>fail)then T:=[];for b in [1..Length(Hz)] do if(b<>a)then Add(T,\mod(Hz[b]+Hz[b][c]*Hz[a],2));fi;od;Hz:=T;fi;fi;;
```

**Comment by qodesign** [comment]
I'm afraid I don't know much about github procedures; you're free to use the code in any way you like but I don't know how involved I can be. One thing I can do is explain the GAP code.

You're right, this is the crux the of calculations. It is based on the steps in the paper : 

The object is to remove one stabilizer and one qubit in each iteration (m is the iteration count and is 1 to the number of secondary qubits = ma*mb).

the "\mod(m,2)" part is to alternate between removing X-stabilizers (rows from Hx) and Z-stabilizers. The paper says this results in the remaining stabilizers having lower row weights. In principle you could have removed any combination of X and Z that add up to ma*mb...I tried a few and that didn't seem to affect the reduced code parameters.

"Ind" is list of secondary qubits to be removed; "c" is the first element in it.
 a:=First([1..Length(Hz)],x->Hz[x][c]=1); finds the first row in Hz that 1 in column "c"; note that Hz[b][c]=1 iff qubit c is connected to Z-stabilizer b. The paper says to drop stabilizer and add it to the other ones in the column.
That's what this  for b in [1..Length(Hz)] do if(b<>a)then Add(T,\mod(Hz[b]+Hz[b][c]*Hz[a],2)); is doing.
So if a=b the row is skipped, otherwise if Hz[b][c]=1 then Hz[a] is added to it and if Hz[b][c] is left alone (not involved in this qubit).
The "fail" condition is in case the column has no 1's at all. This happens at later iterations and seems only necessary for the toric code.
I don't have an explanation for this yet...probably related to the toric code stabilizers being redundant.

After that both Hx and Hz are adjusted because there's one less qubit; the index of the remaining secondary qubits is also adjusted.
Then the loop iterates with the new Hx and Hz...

**Comment by perlinm** [comment]
Rather than removing qubits incrementally, would it be possible to just keep a list of qubits to drop at the end?

Also, why is it guaranteed that removing one stabilizer and one qubit in this way preserves code distance?

**Comment by qodesign** [comment]
it probably is possible to keep the qubits until the end and drop them...dropping was more natural (to me) because I can check the code parameters at every step of the iterations.

The paper explains why this is guaranteed and through first reading it seems reasonable. This is very specific to HGP codes and the way the left and right qubits ("primary" and "secondary") are related.

I did see that for the toric code of odd dimension, distance is not preserved...I don't have an explanation for that yet : could be a bug in my code...or could be a complication because of redundancy in stabilizers

**Comment by perlinm** [comment]
Interesting.  Alright thanks again!  Sketching a (possibly incorrect) python implementation for later:
```python
support_x = matrix_x.astype(bool)
for qubit in second_qubit_sector:
    row = np.argmax(support_x[:, qubit])
    matrix_x += matrix_x[row]
matrix_x = matrix_x[:, first_qubit_sector]
```

**Comment by perlinm** [comment]
Changing `HammingCode(3)` to `RingCode(3)`, it looks like this implementation also removes one logical qubit.  So we should figure out what is going on there.

**Comment by qodesign** [comment]
I did see distance not being preserved for odd toric code (RingCode(odd)). I do plan to debug this and have some ideas on what might be happening. For one, the "fail" condition happens if you're trying to remove a qubit that is not involved in Hx...it might still be involved in Hz, so the alternating Hx/Hz removal might be an issue...

Were you able to verify RepetitionCode and the hamming(3) (right dimensions and distance?)

**Comment by perlinm** [comment]
For the `RepetitionCode` and `HammingCode` the current implementation [here](https://github.com/Infleqtion/qLDPC/issues/113#issuecomment-2699345931) preserves distance.  For the `RingCode --> ToricCode`, it does not.

**Comment by qodesign** [comment]
what about RingCode(4) or 6,...
also as experiment, remove only X-stabilizers and see if RingCode(3), 5,7 preserve distance

**Comment by perlinm** [comment]
RingCode(3) does not preserve distance.  If I remove only X-type stabilizers, it does preserve distance, but not logical qubit number.

**Comment by qodesign** [comment]
i get the right number of logicals if i only remove X stabilizers.
...any "experimentation" and "theory" are part of a loop to nail down fine details

**Comment by qodesign** [comment]
I found the notation of that paper hard to follow...it will be interesting to see other ways to get the solution.
Intuitively, removing the support of the stabilizers from the right qubits is not that "special" : it's similar to putting a matrix in lower triangular form. So there should be several ways of doing it by linear algebraic methods. A trickier part is having the result with the lowest row weights; I thought the second paper did this with that target in mind...although there could be ways to get to optimize that part too. 

The code I wrote (bases on second paper) works fine if $H$ is full rank. In this case there are no logical operators on the right qubits and k and d are preserved. The paper doesn't say that specifically as it should have but they do assume it in their explanation of why distance is preserved. It fails if $H$ is not full rank : in that case you do have "right logicals" (your canonical logicals calculations can give you these). This makes sense since if you remove the right qubits then you're removing the right logicals...so no guarantees on k or d. I think this will be a common constraint on all approaches to remove the right qubits. So either the expectation of preserving k and d are removed or the expectation of removing all the right qubits is removed...

Anyway, since the tanner quantum transform is still not clearly defined in my mind; it will be interesting to see if you find a consistent way to manage the goal and expectations of the calculation

**Comment by qodesign** [comment]
I see there's a "quantum Tanner" code construction in release v0.1.0 (QTCode).
Is this related to the Tanner transform above?.

(a minor nitpick while we're on the subject : ToricCode(2) returns duplicate stabilizers )

```
from qldpc import codes
code = codes.ToricCode(2)
print(code)
```

```
ToricCode on 4 qubits
X-type parity checks:
[[1 1 1 1]
 [1 1 1 1]]
Z-type parity checks:
[[1 1 1 1]
 [1 1 1 1]]
```

**Comment by perlinm** [comment]
The QTCode has been around for quite a while 🙂.  Its relationship to the quantum Tanner transformation is not entirely clear to me.  I *believe* that you can recover some QTCodes by applying the quantum Tanner transformation to HGP codes.

Regarding the Toric code... all standard definitions of the Toric code lead to redundant parity checks.  This redundancy is most egregious in the case of ToricCode(2).  Maybe I should just make ToricCode(2) a special case and remove its redundant checks...

**Comment by qodesign** [comment]
Maybe i just noticed QTCode because I searched for "Tanner". The "Toric code" in the package is the "rotated toric code" according to the source code. I think traditionally Toric code is used to refer to the $[[2d^2,2,d]]$ code which can be constructed as an HGP code. The QTCodes construction gives a $[[d^2,2,d]]$ which is different. In particular the d=2 $[[4,2,2]]$ shows up in a number of settings, so I was especially curious to see how you constructed...especially its relation to the standard $[[8,2,2]]$ toric code.

**Comment by perlinm** [comment]
Let's move the conversation to #267 🙂
